### PR TITLE
Implement Dynamodb with RDS fallback

### PR DESCRIPTION
**Pull request by kierenbeckett**
Changes to go with ONSDigital/eq-terraform#135.  Add new dynamodb tables.

### CODE REVIEW

File: global_vars.tf [hunk 1]
```diff
@@ -19,3 +19,31 @@ variable "submitted_responses_write_capacity" {
...
+  default     = 25
+}
+
+variable "eq_session_write_capacity" {
+  description = "The number of Write units for the EQ Session table"
+  default     = 25
+}
+variable "used_jti_claim_read_capacity" {
+  description = "The number of Read units for the Used JTI claims table"
+  default     = 25
```
Comment (jonnyshaw89) [inline global_vars.tf]: This can be low if not nothing as we don't read.
Will it allow `0`?

File: used_jti_claim.tf [hunk 2]
```diff
@@ -0,0 +1,24 @@
+resource "aws_dynamodb_table" "used_jti_claim_table" {
```
Comment (jonnyshaw89) [inline used_jti_claim.tf:1]: We should add TTL to this table

**Review by jonnyshaw89 — CHANGES_REQUESTED** [review]

File: global_vars.tf (hunk 1 above)
Comment (kierenbeckett) [reply global_vars.tf]: Won't allow 0, have set it to 1

File: eq_session.tf [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+  lifecycle {
+    ignore_changes = ["read_capacity", "write_capacity"]
+  }
+}
+
+resource "aws_appautoscaling_target" "eq_session_table_read_target" {
+  max_capacity       = 100
+  min_capacity       = ${var.eq_session_read_capacity}
+  resource_id        = "table/${var.env}-eq-session"
```
Comment (jonnyshaw89) [inline eq_session.tf]: can this not come from the `aws_dynamodb_table` resource?

File: eq_session.tf [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+    predefined_metric_specification {
+      predefined_metric_type = "DynamoDBReadCapacityUtilization"
+    }
+
+    target_value = 70
+  }
+}
+
+resource "aws_cloudwatch_metric_alarm" "eq_session_table_read_throughput" {
+  alarm_name = "${var.env}-dynamo-eq-session-read-throughput"
```
Comment (jonnyshaw89) [inline eq_session.tf]: `dynamodb`

File: eq_session.tf [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+  comparison_operator = "GreaterThanOrEqualToThreshold"
+  evaluation_periods = "2"
```
Comment (jonnyshaw89) [inline eq_session.tf]: Is 2 minutes too late for the alarm?

File: eq_session.tf [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+  metric_name = "ConsumedReadCapacityUnits"
+  namespace = "AWS/DynamoDB"
+  period = "60"
+  statistic = "Average"
+  threshold = "${var.eq_session_read_capacity * 0.8}"
+  alarm_description = "Average EQ Session DynamoDB read throughput above Threshold of 80% for past 120 seconds"
+  alarm_actions       = ["arn:aws:sns:eu-west-1:${data.aws_caller_identity.current.account_id}:${var.env}-slack-alert"]
```
Comment (jonnyshaw89) [inline eq_session.tf]: Pass the ARN into the module

File: eq_session.tf [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+}
+
+resource "aws_cloudwatch_metric_alarm" "eq_session_table_read_throughput" {
+  alarm_name = "${var.env}-dynamo-eq-session-read-throughput"
+  comparison_operator = "GreaterThanOrEqualToThreshold"
+  evaluation_periods = "2"
+  metric_name = "ConsumedReadCapacityUnits"
+  namespace = "AWS/DynamoDB"
+  period = "60"
+  statistic = "Average"
```
Comment (jonnyshaw89) [inline eq_session.tf]: Maybe `Maximum` so we get the warning earlier?

**Review by jonnyshaw89 — CHANGES_REQUESTED** [review]

File: eq_session.tf (hunk 3 above)
Comment (kierenbeckett) [reply eq_session.tf]: Yup, was still working on it

File: eq_session.tf (hunk 7 above)
Comment (kierenbeckett) [reply eq_session.tf]: This metric is actually the per request consumed capacity, I had to change it to `Sum`

File: eq_session.tf (hunk 5 above)
Comment (kierenbeckett) [reply eq_session.tf]: I was torn, I've known these alarms to be a bit jumpy to temporary spikes in traffic.  Maybe lets start it at one minute and tune it if it gets noisy?

File: eq_session.tf (hunk 6 above)
Comment (kierenbeckett) [reply eq_session.tf]: I based this off the pattern in `survey-runner-database` and `survey-runner-queue`.  Would you prefer passing in the ARN to it being consistent? (or should I refactor the others too?)

Comment (jonnyshaw89) [reply eq_session.tf]: i think passing the ARN is the preference like we have started doing with other resources.
might as well refactor as its the same repo

File: eq_session.tf (hunk 7 above)
Comment (jonnyshaw89) [reply eq_session.tf]: ah ok, if its per request then the period will have a huge impact.

File: global_vars.tf [hunk 8]
```diff
@@ -10,12 +10,82 @@ variable "aws_access_key" {
...
+variable "submitted_responses_min_read_capacity" {
+  description = "The min number of Read units for the Submitted Responses table"
   default     = 25
 }
 
-variable "submitted_responses_write_capacity" {
-  description = "The number of Write units for the Submitted Responses table"
+variable "submitted_responses_max_read_capacity" {
+  description = "The max number of Read units for the Submitted Responses table"
+  default     = 500
```
Comment (jonnyshaw89) [inline global_vars.tf]: I would set these a lot higher as they are production default

Comment (jonnyshaw89) [reply global_vars.tf]: 5000?
default account limit is 20,000 read 20,000 write so will give even split across tables.
we can tune and request more once its in

**Review by jonnyshaw89 — CHANGES_REQUESTED** [review]

**Review by nealedj — APPROVED** [review]

**Review by jonnyshaw89 — APPROVED** [review]

**Comment by jonnyshaw89** [comment]
@kierenbeckett can you squash this branch?

File: questionnaire_state.tf [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+resource "aws_cloudwatch_metric_alarm" "questionnaire_state_table_read_throughput" {
+  alarm_name          = "${var.env}-dynamodb-questionnaire-state-read-throughput"
+  comparison_operator = "GreaterThanOrEqualToThreshold"
+  evaluation_periods  = "1"
+  metric_name         = "ConsumedReadCapacityUnits"
+  namespace           = "AWS/DynamoDB"
+  period              = "60"
+  statistic           = "Sum"
+  threshold           = "${var.questionnaire_state_max_read_capacity * 0.8 * 60}"
+  alarm_description   = "Questionnaire State DynamoDB average consumed read throughput above 80% of maximum for past 60 seconds"
```
Comment (jonnyshaw89) [inline questionnaire_state.tf]: You should prefix these alarms with EQ as per the session alarms. as in preprod and prod they go to an SDC wide ops channel

File: submitted_responses.tf [hunk 10]
```diff
@@ -18,6 +18,160 @@ resource "aws_dynamodb_table" "submitted_responses_table" {
...
+resource "aws_cloudwatch_metric_alarm" "submitted_responses_table_read_throughput" {
+  alarm_name          = "${var.env}-dynamodb-submitted-responses-read-throughput"
+  comparison_operator = "GreaterThanOrEqualToThreshold"
+  evaluation_periods  = "1"
+  metric_name         = "ConsumedReadCapacityUnits"
+  namespace           = "AWS/DynamoDB"
+  period              = "60"
+  statistic           = "Sum"
+  threshold           = "${var.submitted_responses_max_read_capacity * 0.8 * 60}"
+  alarm_description   = "Submitted Responses DynamoDB average consumed read throughput above 80% of maximum for past 60 seconds"
```
Comment (jonnyshaw89) [inline submitted_responses.tf]: Same comment re EQ prefix

File: used_jti_claim.tf [hunk 11]
```diff
@@ -0,0 +1,183 @@
...
+resource "aws_cloudwatch_metric_alarm" "used_jti_claim_table_read_throughput" {
+  alarm_name          = "${var.env}-dynamodb-used-jti-claim-read-throughput"
+  comparison_operator = "GreaterThanOrEqualToThreshold"
+  evaluation_periods  = "1"
+  metric_name         = "ConsumedReadCapacityUnits"
+  namespace           = "AWS/DynamoDB"
+  period              = "60"
+  statistic           = "Sum"
+  threshold           = "${var.used_jti_claim_max_read_capacity * 0.8 * 60}"
+  alarm_description   = "Used JTI Claim DynamoDB average consumed read throughput above 80% of maximum for past 60 seconds"
```
Comment (jonnyshaw89) [inline used_jti_claim.tf]: Here too

**Review by jonnyshaw89 — CHANGES_REQUESTED** [review]
Feel free to squash with this change and then we can get this merged

**Review by jonnyshaw89 — APPROVED** [review]

**Review by jonnyshaw89 — CHANGES_REQUESTED** [review]
1 final change. The README is now out of date

File: README.md [hunk 12]
```diff
@@ -21,6 +36,21 @@ To run this module on its own run the following code: (Replacing 'XXX' with your
...
+                -var "slack_alert_sns_arn=XXX" \
+                -var "submitted_responses_min_read_capacity=XXX" \
+                -var "submitted_responses_max_read_capacity=XXX" \
+                -var "submitted_responses_min_write_capacity=XXX" \
+                -var "submitted_responses_max_write_capacity=XXX" \
+                -var "questionnaire_state_min_read_capacity=XXX" \
+                -var "questionnaire_state_max_read_capacity=XXX" \
+                -var "questionnaire_state_min_write_capacity=XXX" \
+                -var "questionnaire_state_max_write_capacity=XXX" \
+                -var "eq_sessio_min_read_capacity=XXX" \
```
Comment (jonnyshaw89) [inline README.md]: typo `eq_sessio`

**Review by jonnyshaw89 — APPROVED** [review]

**Pull request merged**